src/strgen/strgen.cpp

```diff
@@ -485,78 +396,74 @@ int CDECL main(int argc, char *argv[])
 				return 0;
 
 			case 's':
-				src_dir = replace_pathsep(mgo.opt);
+				src_dir = mgo.opt;
 				break;
 
 			case 'd':
-				dest_dir = replace_pathsep(mgo.opt);
+				dest_dir = mgo.opt;
 				break;
 
 			case -2:
-				fprintf(stderr, "Invalid arguments\n");
+				fmt::print(stderr, "Invalid arguments\n");
 				return 0;
 		}
 	}
 
-	if (dest_dir == nullptr) dest_dir = src_dir; // if dest_dir is not specified, it equals src_dir
+	if (dest_dir.empty()) dest_dir = src_dir; // if dest_dir is not specified, it equals src_dir
 
 	try {
 		/* strgen has two modes of operation. If no (free) arguments are passed
 		 * strgen generates strings.h to the destination directory. If it is supplied
 		 * with a (free) parameter the program will translate that language to destination
 		 * directory. As input english.txt is parsed from the source directory */
 		if (mgo.numleft == 0) {
-			mkpath(pathbuf, lastof(pathbuf), src_dir, "english.txt");
+			std::filesystem::path input_path = src_dir;
+			input_path /= "english.txt";
 
 			/* parse master file */
 			StringData data(TEXT_TAB_END);
-			FileStringReader master_reader(data, pathbuf, true, false);
+			FileStringReader master_reader(data, input_path, true, false);
 			master_reader.ParseFile();
 			if (_errors != 0) return 1;
 
 			/* write strings.h */
-			ottd_mkdir(dest_dir);
-			mkpath(pathbuf, lastof(pathbuf), dest_dir, "strings.h");
+			std::filesystem::path output_path = dest_dir;
+			std::filesystem::create_directories(dest_dir);
+			output_path /= "strings.h";
 
-			HeaderFileWriter writer(pathbuf);
+			HeaderFileWriter writer(output_path);
 			writer.WriteHeader(data);
 			writer.Finalise(data);
 			if (_errors != 0) return 1;
 		} else if (mgo.numleft >= 1) {
-			char *r;
-
-			mkpath(pathbuf, lastof(pathbuf), src_dir, "english.txt");
+			std::filesystem::path input_path = src_dir;
+			input_path /= "english.txt";
 
 			StringData data(TEXT_TAB_END);
 			/* parse master file and check if target file is correct */
-			FileStringReader master_reader(data, pathbuf, true, false);
+			FileStringReader master_reader(data, input_path, true, false);
 			master_reader.ParseFile();
 
 			for (int i = 0; i < mgo.numleft; i++) {
 				data.FreeTranslation();
 
-				const char *translation = replace_pathsep(mgo.argv[i]);
-				const char *file = strrchr(translation, PATHSEPCHAR);
-				FileStringReader translation_reader(data, translation, false, file == nullptr || strcmp(file + 1, "english.txt") != 0);
+				std::filesystem::path lang_file = mgo.argv[i];
+				FileStringReader translation_reader(data, lang_file, false, lang_file.filename() != "english.txt");
```

Comment:
The whole idea of strgen is that we can inject paths as command-line arguments.

Reply:
isn't the issue with these that access is unchecked? I realise it's unnecessary for our purposes, but perhaps some basic validity checks would make it go away? https://codeql.github.com/codeql-query-help/cpp/cpp-path-injection/#recommendation

Reply:
From the suggestions:
- Do not allow more than a single “.” character.
  - Given CMake uses full paths, so checking for this means a builder is not allowed to have a `.` in the build path. I do not think we should add this limitation.
- Do not allow directory separators such as “/” or “\” (depending on the filesystem).
  - Given CMake uses full paths, this breaks building with CMake.
- Do not rely on simply replacing problematic sequences such as “../”. For example, after applying this filter to “…/…//” the resulting string would still be “../”.
  - There is no reliance on removing "../". Doesn't seem very useful to me, as then you can't easily run strgen without providing the full path which would be a nuisance.
- Ideally use a whitelist of known good patterns.
  - Can you provide me with a list of the paths of all the builders? Or are we going to "just" check that it contains "lang" as whitelist; the latter seems pretty pointless.

So, what check can I implement? Just checking for "lang" in the filename seems very pointless and that yields the question... why? And the answer: to silence a code scanner on a warning that we will not comply to and we tried to silence, instead of just dismissing it with a proper comment.